GVFS/GVFS.Virtualization/FileSystem/FileSystemVirtualizer.cs

```diff
@@ -208,6 +208,62 @@ protected void OnWorkingDirectoryFileOrFolderDeleted(string relativePath, bool i
 
             this.FileSystemCallbacks.InvalidateGitStatusCache();
         }
+        
+        protected void OnFileRenamed(string relativeSourcePath, string relativeDestinationPath, bool isDirectory)
+        {
+            try
+            {
+                bool srcPathInDotGit = FileSystemCallbacks.IsPathInsideDotGit(relativeSourcePath);
+                bool dstPathInDotGit = FileSystemCallbacks.IsPathInsideDotGit(relativeDestinationPath);
+
+                if (dstPathInDotGit)
+                {
+                    this.OnDotGitFileOrFolderChanged(relativeDestinationPath);
+                }
+
+                if (!(srcPathInDotGit && dstPathInDotGit))
+                {
+                    if (isDirectory)
+                    {
+                        this.FileSystemCallbacks.OnFolderRenamed(relativeSourcePath, relativeDestinationPath);
+                    }
+                    else
+                    {
+                        this.FileSystemCallbacks.OnFileRenamed(relativeSourcePath, relativeDestinationPath);
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                EventMetadata metadata = this.CreateEventMetadata(relativeSourcePath, e);
+                metadata.Add("destinationPath", relativeDestinationPath);
+                metadata.Add("isDirectory", isDirectory);
+                this.LogUnhandledExceptionAndExit(nameof(this.OnFileRenamed), metadata);
+            }
+        }
+
+        protected void OnHardLinkCreated(string relativeTargetPath, string relativeNewLinkPath)

```

Comment:
Why do we even need the `relativeTargetPath` parameter? It doesn't seem to be used at all.

Reply:
And why would it always be relative? Couldn't you create a hardlink inside your repo to some path outside of it? It seems like the target would need to be treated as an absolute path.

Reply:
> Why do we even need the relativeTargetPath parameter? It doesn't seem to be used at all.

It's not *strictly* needed, but we do log it if there's an exception:

`metadata.Add(nameof(relativeTargetPath), relativeTargetPath);`

My thought was that if something goes wrong it's helpful to have as much information as we can.

> And why would it always be relative? Couldn't you create a hardlink inside your repo to some path outside of it? It seems like the target would need to be treated as an absolute path.

See related comment as well (for detailed descriptions of these paths), if either of the paths are outside of the repo they will be empty strings.